src/licensedcode/detection.py

```diff
@@ -295,19 +673,154 @@ def is_license_intro(license_match):
     )
 
 
-class CombinationReason(Enum):
-    NOT_COMBINED = 'not-combined'
-    UNKNOWN_INTRO_FOLLOWED_BY_MATCH = 'unknown-intro-followed-by-match'
-    CONTAINED_SAME_LICENSE = 'contained-with-same-license'
-    NOTICE_FOLLOWED_BY_TEXT = 'notice-followed-by-text'
-    CONTIGUOUS_SAME_LICENSE = 'contiguous-with-same-license'
-    REF_FOLLOWED_BY_NOTICE = 'ref-followed-by-notice'
-    REF_FOLLOWED_BY_TEXT = 'ref-followed-by-text'
-    TAG_FOLLOWED_BY_NOTICE = 'tag-followed-by-notice'
-    TAG_FOLLOWED_BY_TEXT = 'tag-followed-by-text'
-    TAG_FOLLOWED_BY_REF = 'tag-followed-by-ref'
-    UNKNOWN_FOLLOWED_BY_MATCH = 'unknown-ref-followed-by-match'
-    UNVERSIONED_FOLLOWED_BY_VERSIONED = 'un-versioned-followed-by-versioned'
+def is_license_reference_local_file(license_match):
+    """
+    Return True if `license_match` LicenseMatch dict has a
+    non-empty `referenced_filename`, i.e. contains a license
+    reference to a local file.
+    """
+    if license_match['matched_rule']['referenced_filenames']:
+        return True
+
+    return False
+
+
+def combine_license_references(license_matches):
+    """
+    Return a filtered ``license_matches`` list of LicenseMatch objects removing
+    references to local files with licenses.
+
+    A common source of unknown license detections in license references to
+    a local file like "see license in LICENSE.md". In these cases we can
+    remove the unknown reference with the license detection in the file.
+    """
+    return [match for match in license_matches if not is_license_reference_local_file(match)]
+
+
+def get_detected_license_expression(matches, analysis):
+    """
+    Get `reasons` and `combined_license_expression` after combining
+    `matches` list of LicenseMatch objects.
+    """
+    matches_for_expression = None
+    reasons = []
+
+    if analysis == "unknown-intro-before-detection":
+        matches_for_expression = combine_license_intros(matches)
+        reasons.append(CombinationReason.UNKNOWN_INTRO_FOLLOWED_BY_MATCH.value)
+    elif analysis == "unknown-file-reference-local":
+        matches_for_expression = combine_license_references(matches)
+        reasons.append(CombinationReason.UNKNOWN_REFERENCE_TO_LOCAL_FILE.value)
+    else:
+        matches_for_expression = matches
+
+    if isinstance(matches[0], dict):
+        combined_expression = combine_expressions(
+            expressions=[match['matched_rule']['license_expression'] for match in matches_for_expression]
```

Comment:
Should you rather use deserialized real LicenseMatch objects throughout and rules recreated from the scan data?
This would make the whole code cleaner rather than using dicts and key lookups.

Reply:
This would make the code at process_codebase much more complicated tho, as we use have to serialize and de-serialize all the detections (and matches in them?)